File: frontend/src/app/core/services/raw-material-inventory.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, map } from 'rxjs';
import { environment } from '../../../environments/environment';
import {
  WarehouseLocation, GRN, StockLot, StockLotDetail,
  StockSummary, Adjustment,
} from '../models/raw-material-inventory.model';

@Injectable({ providedIn: 'root' })
export class RawMaterialInventoryService {
  private http = inject(HttpClient);
  private base = `${environment.apiUrl}/inventory/raw-material`;

  listWarehouseLocations(): Observable<WarehouseLocation[]> {
    return this.http.get<{ data: WarehouseLocation[] }>(`${this.base}/warehouse-locations`).pipe(map(r => r.data));
  }

  // GRN
  listGRNs(): Observable<GRN[]> {
    return this.http.get<{ data: GRN[] }>(`${this.base}/grn`).pipe(map(r => r.data));
  }

  getGRN(id: string): Observable<GRN> {
    return this.http.get<GRN>(`${this.base}/grn/${id}`);
  }

  createGRN(body: object): Observable<GRN> {
    return this.http.post<GRN>(`${this.base}/grn`, body);
  }

  updateGRN(id: string, body: object): Observable<GRN> {
    return this.http.put<GRN>(`${this.base}/grn/${id}`, body);
  }

  confirmGRN(id: string): Observable<GRN> {
    return this.http.post<GRN>(`${this.base}/grn/${id}/confirm`, {});
  }

  cancelGRN(id: string): Observable<void> {
    return this.http.post<void>(`${this.base}/grn/${id}/cancel`, {});
  }

  // Stock
  listStockSummary(): Observable<StockSummary[]> {
    return this.http.get<{ data: StockSummary[] }>(`${this.base}/stock/summary`).pipe(map(r => r.data));
  }

  listStockLots(rawMaterialId?: string): Observable<StockLot[]> {
    const params = rawMaterialId ? `?raw_material_id=${rawMaterialId}` : '';
    return this.http.get<{ data: StockLot[] }>(`${this.base}/stock/lots${params}`).pipe(map(r => r.data));
  }

  getStockLotDetail(id: string): Observable<StockLotDetail> {
    return this.http.get<StockLotDetail>(`${this.base}/stock/lots/${id}`);
  }

  // Adjustments
  listAdjustments(): Observable<Adjustment[]> {
    return this.http.get<{ data: Adjustment[] }>(`${this.base}/adjustments`).pipe(map(r => r.data));
  }

  getAdjustment(id: string): Observable<Adjustment> {
    return this.http.get<Adjustment>(`${this.base}/adjustments/${id}`);
  }

  createAdjustment(body: object): Observable<Adjustment> {
    return this.http.post<Adjustment>(`${this.base}/adjustments`, body);
  }

  approveAdjustment(id: string): Observable<Adjustment> {
    return this.http.post<Adjustment>(`${this.base}/adjustments/${id}/approve`, {});
  }

  cancelAdjustment(id: string): Observable<void> {
    return this.http.post<void>(`${this.base}/adjustments/${id}/cancel`, {});
  }
}

File: backend/internal/service/raw_material_inventory_service.go
package service

import (
	"context"

	"github.com/water-factory/api/internal/domain"
	"github.com/water-factory/api/internal/dto"
	"github.com/water-factory/api/internal/repository"
)

type RMInventoryService struct {
	repo *repository.RMInventoryRepo
}

func NewRMInventoryService(repo *repository.RMInventoryRepo) *RMInventoryService {
	return &RMInventoryService{repo: repo}
}

func (s *RMInventoryService) ListWarehouseLocations(ctx context.Context) ([]*domain.WarehouseLocation, error) {
	return s.repo.ListWarehouseLocations(ctx)
}

func (s *RMInventoryService) ListGRNs(ctx context.Context) ([]*domain.GRN, error) {
	return s.repo.ListGRNs(ctx)
}

func (s *RMInventoryService) GetGRN(ctx context.Context, id string) (*domain.GRN, error) {
	return s.repo.GetGRN(ctx, id)
}

func (s *RMInventoryService) CreateGRN(ctx context.Context, userID string, req *dto.CreateGRNRequest) (*domain.GRN, error) {
	return s.repo.CreateGRN(ctx, userID, req)
}

func (s *RMInventoryService) UpdateGRN(ctx context.Context, id string, req *dto.UpdateGRNRequest) (*domain.GRN, error) {
	return s.repo.UpdateGRN(ctx, id, req)
}

func (s *RMInventoryService) ConfirmGRN(ctx context.Context, id, userID string) (*domain.GRN, error) {
	return s.repo.ConfirmGRN(ctx, id, userID)
}

func (s *RMInventoryService) CancelGRN(ctx context.Context, id string) error {
	return s.repo.CancelGRN(ctx, id)
}

func (s *RMInventoryService) ListStockSummary(ctx context.Context) ([]*domain.StockSummary, error) {
	return s.repo.ListStockSummary(ctx)
}

func (s *RMInventoryService) ListStockLots(ctx context.Context, rawMaterialID string) ([]*domain.StockLot, error) {
	return s.repo.ListStockLots(ctx, rawMaterialID)
}

func (s *RMInventoryService) GetStockLotDetail(ctx context.Context, id string) (*domain.StockLotDetail, error) {
	return s.repo.GetStockLotDetail(ctx, id)
}

func (s *RMInventoryService) ListAdjustments(ctx context.Context) ([]*domain.Adjustment, error) {
	return s.repo.ListAdjustments(ctx)
}

func (s *RMInventoryService) GetAdjustment(ctx context.Context, id string) (*domain.Adjustment, error) {
	return s.repo.GetAdjustment(ctx, id)
}

func (s *RMInventoryService) CreateAdjustment(ctx context.Context, userID string, req *dto.CreateAdjustmentRequest) (*domain.Adjustment, error) {
	return s.repo.CreateAdjustment(ctx, userID, req)
}

func (s *RMInventoryService) ApproveAdjustment(ctx context.Context, id, approverID string) (*domain.Adjustment, error) {
	return s.repo.ApproveAdjustment(ctx, id, approverID)
}

func (s *RMInventoryService) CancelAdjustment(ctx context.Context, id string) error {
	return s.repo.CancelAdjustment(ctx, id)
}
